sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizedLocalShuffleReader.scala

```diff
@@ -0,0 +1,124 @@
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.physical.{Partitioning, UnknownPartitioning}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.{SparkPlan, UnaryExecNode}
+import org.apache.spark.sql.execution.exchange.{EnsureRequirements, ShuffleExchangeExec}
+import org.apache.spark.sql.execution.joins.{BroadcastHashJoinExec}
+import org.apache.spark.sql.internal.SQLConf
+
+case class OptimizedLocalShuffleReader(conf: SQLConf) extends Rule[SparkPlan] {
+
+  private def setIsLocalToFalse(shuffleStage: QueryStageExec): QueryStageExec = {
+    shuffleStage match {
+      case stage: ShuffleQueryStageExec =>
+        stage.isLocalShuffle = false
+      case ReusedQueryStageExec(_, stage: ShuffleQueryStageExec, _) =>
+        stage.isLocalShuffle = false
+    }
+    shuffleStage
+  }
+
+  private def revertLocalShuffleReader(newPlan: SparkPlan): SparkPlan = {
+    val revertPlan = newPlan.transformUp {
+      case localReader: LocalShuffleReaderExec
+        if (ShuffleQueryStageExec.isShuffleQueryStageExec(localReader.child)) =>
+        setIsLocalToFalse(localReader.child)
+    }
+    revertPlan
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    // Collect the `BroadcastHashJoinExec` nodes and if isEmpty directly return.
+    val bhjs = plan.collect {
+      case bhj: BroadcastHashJoinExec => bhj
+    }
+
+    if (!conf.optimizedLocalShuffleReaderEnabled || bhjs.isEmpty) {
+      return plan
+    }
+
+    // If the streamedPlan is `ShuffleQueryStageExec`, set the value of `isLocalShuffle` to true
+    bhjs.map {
+      case bhj: BroadcastHashJoinExec =>
+        bhj.children map {
+          case stage: ShuffleQueryStageExec => stage.isLocalShuffle = true
+          case ReusedQueryStageExec(_, stage: ShuffleQueryStageExec, _) =>
+            stage.isLocalShuffle = true
+          case plan: SparkPlan => plan
+        }
+    }
+
+    // Add the new `LocalShuffleReaderExec` node if the value of `isLocalShuffle` is true
+    val newPlan = plan.transformUp {
+      case stage: ShuffleQueryStageExec if (stage.isLocalShuffle) =>
+        LocalShuffleReaderExec(stage)
+      case ReusedQueryStageExec(_, stage: ShuffleQueryStageExec, _) if (stage.isLocalShuffle) =>
+        LocalShuffleReaderExec(stage)
+    }
+
+    val afterEnsureRequirements = EnsureRequirements(conf).apply(newPlan)
+    val numExchanges = afterEnsureRequirements.collect {
+      case e: ShuffleExchangeExec => e
+    }.length
+    if (numExchanges > 0) {
+      logWarning("Local shuffle reader optimization is not applied due" +
+        " to additional shuffles will be introduced.")
+      revertLocalShuffleReader(newPlan)
+    } else {
+      newPlan
+    }
+  }
+}
+
+case class LocalShuffleReaderExec(
+    child: QueryStageExec) extends UnaryExecNode {
```

Comment:
We can make it a leaf node to hide its `QueryStageExec`. We don't expect any other rules to change the underlying shuffle stage.

Reply:
Good suggestions. Updated.